### ISSUE TITLE
Turning ae_mdw into an aeternity node plugin

### ISSUE THREAD

**Issue by sborrazas** [aeternity/ae_mdw#334]
## Current state

To provide some context, currently this middleware application works by loading the aeternity node compiled erlang modules. This is done by [ae_plugin](https://github.com/aeternity/ae_plugin/), which is loaded when the mdw starts, following these steps:

1. The `:ae_plugin, :ae_root` variable is used to decide where the node compiled release root is
2. Using that directory as base, all of the `<root>/rel/aeternity/lib/*/ebin` paths are added to the erlang loading paths
3. Using that directory as base, all of the `<root>/rel/aeternity/releases/*/sys.config` are loaded into the env
4. The `aecore` app is then started

 In comparison, the [devmode plugin](https://github.com/aeternity/aeplugin_dev_mode) does a completely different thing: 
1. Once `aecore` starts it checks whether there are any plugins listed in the config. 
2. If there are any, it will try to load them by looking into the local `plugins` directory, loading the modules and starting the plugin.

In this last case, the plugin dependencies need to be already loaded. In the case of devmode, all of the dependencies are already loaded in the core app, so it doesn't need to load anything else.

## Future thoughts

One idea to make the middleware plugin-compatible would be to have the middleware start as an application that would only load all of the dependancies needed, similar to what the `ae_plugin` does. This would involve a lot of work and would make the elixir app not the main application started, which would make it rather hard for it to act like a regular Elixir app. Not only that, but it would also need to load and start all of the Elixir modules that are part of the Elixir stdlib.

Another idea would be for the middleware to communicate with the core application through EDTS and RPC (erlang messages). This would involve a lot of work as well.

For now, the middleware application will remain behaving like it does: started as a regular Elixir app and loading the core application libs and deps on startup. The best way for users to configure and take advantage of the middleware is to use the Docker image provided in this repo.

Any other ideas, comments or suggestions are more than welcomed!

**Comment by marc0olo** [comment]
Thanks for the analysis @sborrazas. This sounds like we need to stick to the current behavior unless we are willing to put lots of efforts into this, right?

Is there any feasible way to provide an improved docker build (https://github.com/aeternity/ae_mdw/issues/195) that allows us to publish the latest image after each release? ideally 2 different images:
- the bundle of the latest node release (where the plugins are included)
- the latest node release including mdw

I am not sure if there is lots of demand for that. but I personally really want to have a well defined process and published images in that area.

**Comment by thepiwo** [comment]
@marc0olo what will be the goal of 2 images from mdw side?

**Comment by marc0olo** [comment]
> @marc0olo what will be the goal of 2 images from mdw side?

not sure. just thinking. maybe somebody doesn't want to have devmode (and other plugins) included which come from the release bundle of the node

**Comment by marc0olo** [comment]
if we decide to not support /publish a mdw image that includes other plugins and specifically devmode we should at least provide a script that combines this so that people can build the bundle image by themselves easily.

just my personal opinion here.

**Comment by thepiwo** [comment]
I don't think the limiting factor is a docker image here, but compatibility of devmode and mdw

**Comment by marc0olo** [comment]
ok maybe I didn't read the analysis deep enough. if I understand you correctly you say that it's generally not easily possible to build and run an application that includes both, devmode and mdw???

I thought it was "just" about making mdw to use the same mechanism like devmode does.

**Comment by thepiwo** [comment]
yes its not easily possible, devmode uses some hacks in the node, e.g. to rollback that are not compatible with mdw in its current state

**Comment by marc0olo** [comment]
haha, I think that's quite a tough task for #hacktoberfest @jyeshe - you wanna attract people that like the challenge? :)
